Add QuadTransformerDistributeIri

wvdemeer opened the pull request:
A quad transformer that distributes IRIs over multiple destination IRIs. It does so by interpreting the first capture group as a number and using `replacementStrings[number % replacementStrings.length]`

(credit @twalcari)

rubensworks commented on lib/transform/QuadTransformerDistributeIri.ts:
```diff
@@ -0,0 +1,34 @@
+    if (term.termType === 'NamedNode') {
+      const match = this.search.exec(term.value);
+      if (match) {
+        const nr = Number.parseInt(match[1], 10);
```
Here we assume that the regex will only match numbers, but the user _could_ also provide a regex that matches strings.
I'd suggest adding a try-catch around this to check for number parsing failures, and re-throw a new error saying that the user provided an unexpected regex that matches non-numbers.
Additionally, we should also document this number-matching requirement on the regex.

rubensworks reviewed (changes requested):
Thanks, looks great!

Could you also add a brief README entry on this with an example and description of how this new transformer can be used?

wvdemeer commented:
Added info to `README.md`, and added clear errors to `QuadTransformerDistributeIri.transformTerm`.
Let me know if there's anything else that needs to be changed.

rubensworks commented on lib/transform/QuadTransformerDistributeIri.ts:
```diff
@@ -0,0 +1,46 @@
+      const match = this.search.exec(term.value);
+      if (match) {
+        if (match.length < 2) {
+          throw new Error(`QuadTransformerDistributeIri error: The "searchRegex" did not contain any groups. ` +
+              `QuadTransformerDistributeIri requires at least one group-based replacement, ` +
+              `of which the first group must match a number.`);
```
```suggestion
          throw new Error(`'searchRegex' did not contain any groups, while QuadTransformerDistributeIri requires at least one group-based replacement, of which the first group must match a number.`);
```

rubensworks commented on lib/transform/QuadTransformerDistributeIri.ts:
```diff
@@ -0,0 +1,44 @@
+        }
+        const nr = Number.parseInt(match[1], 10);
+        if (Number.isNaN(nr)) {
+          throw new Error(`QuadTransformerDistributeIri error: The first capture group in "searchRegex"` +
+              ` must always match a number, but it matched "${match[1]}" instead.`);
```
```suggestion
          throw new Error(`The first capture group in 'searchRegex' must always match a number, but it matched "${match[1]}" instead.`);
```

rubensworks commented:
Thanks for the update!

Could you add unit tests for those two error cases? (The CI required 100% test coverage)

wvdemeer commented:
100% test coverage now.

rubensworks commented:
Nice, thanks @wvdemeer!
